pilot/pkg/bootstrap/server.go

```diff
@@ -591,29 +576,64 @@ func (s *Server) createK8sServiceControllers(serviceControllers *aggregate.Contr
 	return
 }
 
-// initMultiClusterController initializes multi cluster controller
-// currently implemented only for kubernetes registries
-func (s *Server) initMultiClusterController(args *PilotArgs) error {
-	if hasKubeRegistry(args) {
-		s.addStartFunc(func(stop <-chan struct{}) error {
-			secretController := clusterregistry.NewController(s.kubeClient,
-				args.Config.ClusterRegistriesNamespace,
-				s.clusterStore,
-				s.ServiceController,
-				s.EnvoyXdsServer,
-				args.Config.ControllerOptions.ResyncPeriod,
-				args.Config.ControllerOptions.WatchedNamespace,
-				args.Config.ControllerOptions.DomainSuffix)
-
-			// Start secret controller which watches for runtime secret Object changes and adds secrets dynamically
-			go secretController.Run(stop)
+// initMultiClusterSecretController initializes multi cluster secret controller
+// to watch for remote cluster additions or deletions.
+// Currently implemented only for kubernetes registries
+func (s *Server) initMultiClusterSecretController(args *PilotArgs) (err error) {
+	if checkForKubernetes(args.Service.Registries) {
+
+		err = secretcontroller.StartSecretController(s.kubeClient, s.addMemberCluster, s.deleteMemberCluster, args.Config.ClusterRegistriesNamespace)
```

Comment:
I think here should revert what #8511 does: pass a stop channel.

Reply:
Ok - will look in to changes required.

Reply:
After looking at this more I believe the implementation I have is correct and the changes that are in 8511 should be applied against the common secret controller library not this code.   This code creates a controller when a cluster is added.  When that cluster is deleted we would like to stop all the go routines that controller is running.  We need to send a stop to do that.   This is independent of the stopping when the main pilot server stops.

Reply:
@john-a-joyce When pilot stops send a notification to all controllers started by `secret controller`, which is a grace shutdown way.